File: LARGERslicer/Utils/ExportNamingHelper.cs
using System;
using System.IO;
using System.Text;

namespace LARGERslicer.Utils
{
    internal static class ExportNamingHelper
    {
        public static string BuildBaseName(
            DateTime date,
            string projectName,
            string partNumber,
            string subPart,
            string revision,
            string documentType)
        {
            string prefix = "PS" + date.ToString("yyyyMMdd");
            string project = SanitizeToken(projectName, "Projekt");
            string part = SanitizeToken(partNumber, "Teil");
            string sub = SanitizeToken(subPart, "Allgemein");
            string doc = SanitizeToken(documentType, "Geometrie");
            string rev = SanitizeToken(revision, string.Empty);

            var sb = new StringBuilder();
            sb.Append(prefix);
            sb.Append("_");
            sb.Append(project);
            sb.Append(" - ");
            sb.Append(doc);
            sb.Append(" - ");
            sb.Append(part);
            sb.Append(" - ");
            sb.Append(sub);

            if (!string.IsNullOrWhiteSpace(rev))
            {
                sb.Append(" - ");
                sb.Append(rev);
            }

            return CollapseWhitespace(sb.ToString().Trim());
        }

        public static string EnsureExtension(string extension)
        {
            if (string.IsNullOrWhiteSpace(extension))
                return ".txt";
            return extension.StartsWith(".") ? extension : "." + extension;
        }

        public static string SanitizeToken(string value, string fallback)
        {
            string text = string.IsNullOrWhiteSpace(value) ? fallback : value.Trim();
            var invalid = Path.GetInvalidFileNameChars();
            var sb = new StringBuilder(text.Length);
            foreach (char c in text)
            {
                bool isInvalid = false;
                for (int i = 0; i < invalid.Length; i++)
                {
                    if (c == invalid[i])
                    {
                        isInvalid = true;
                        break;
                    }
                }

                if (isInvalid)
                    sb.Append('_');
                else
                    sb.Append(c);
            }

            return CollapseWhitespace(sb.ToString().Trim());
        }

        private static string CollapseWhitespace(string input)
        {
            if (string.IsNullOrWhiteSpace(input))
                return string.Empty;

            var sb = new StringBuilder(input.Length);
            bool prevWasSpace = false;
            foreach (char c in input)
            {
                bool isSpace = char.IsWhiteSpace(c);
                if (isSpace)
                {
                    if (!prevWasSpace)
                        sb.Append(' ');
                }
                else
                {
                    sb.Append(c);
                }

                prevWasSpace = isSpace;
            }

            return sb.ToString();
        }
    }
}

File: LARGERslicer/documentations/export_naming_helper.py
"""
Export naming helper for GH Python workflows.

Naming schema:
PSYYYYMMDD_Project - DocumentType - PartNumber - SubPart - Revision

Revision is optional and omitted when empty.
"""

import datetime
import System


def collapse_whitespace(text):
    if text is None:
        return ""
    return " ".join(str(text).split())


def sanitize_token(value, fallback=""):
    raw = fallback if value is None or str(value).strip() == "" else str(value).strip()
    invalid = set(System.IO.Path.GetInvalidFileNameChars())

    chars = []
    for c in raw:
        if c in invalid:
            chars.append("_")
        else:
            chars.append(c)

    return collapse_whitespace("".join(chars).strip())


def ensure_extension(ext, default_ext=".txt"):
    if ext is None:
        return default_ext
    value = str(ext).strip()
    if value == "":
        return default_ext
    return value if value.startswith(".") else "." + value


def build_base_name(
    project_name,
    part_number,
    sub_part,
    revision="",
    document_type="Geometrie",
    date_value=None,
):
    if date_value is None:
        date_value = datetime.datetime.now()

    prefix = "PS" + date_value.strftime("%Y%m%d")
    project = sanitize_token(project_name, "Projekt")
    part = sanitize_token(part_number, "Teil")
    sub = sanitize_token(sub_part, "Allgemein")
    doc = sanitize_token(document_type, "Geometrie")
    rev = sanitize_token(revision, "")

    base = "{}_{:s} - {:s} - {:s} - {:s}".format(prefix, project, doc, part, sub)
    if rev:
        base += " - " + rev
    return collapse_whitespace(base.strip())
